core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
     {
         return queryContexts.getUnchecked(queryId);
     }
+
+    @VisibleForTesting
+    public void setStuckSplitStackTracePredicate(Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate)
+    {
+        this.stuckSplitTaskInterrupter.ifPresent(stuckSplitTaskInterrupter ->
+                stuckSplitTaskInterrupter.stuckSplitStackTracePredicate = stuckSplitStackTracePredicate);
+    }
+
+    @VisibleForTesting
+    public void failStuckSplitTasks()
+    {
+        stuckSplitTaskInterrupter.ifPresent(StuckSplitTaskInterrupter::failStuckSplitTasks);
+    }
+
+    private Optional<StuckSplitTaskInterrupter> createStuckSplitTaskInterrupter(
+            boolean enableInterruptStuckSplitTasks,
+            Duration stuckSplitsWarningThreshold,
+            Duration interruptStuckSplitTasksTimeout,
+            Duration stuckSplitsDetectionInterval,
+            Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate,
+            TaskExecutor taskExecutor,
+            Ticker ticker)
+    {
+        if (!enableInterruptStuckSplitTasks) {
+            return Optional.empty();
+        }
+        return Optional.of(new StuckSplitTaskInterrupter(stuckSplitsWarningThreshold, interruptStuckSplitTasksTimeout, stuckSplitsDetectionInterval, stuckSplitStackTracePredicate, taskExecutor, ticker));
+    }
+
+    private class StuckSplitTaskInterrupter
+    {
+        private final Duration interruptStuckSplitTasksTimeout;
+        private final Duration stuckSplitsDetectionInterval;
+        private Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate;
```

Comment:
I would recommend making it final and have a `@VisibleForTesting` version of the `SqlTaskManager` constructor that takes an interrupter that is constructed in a certain way (with a predicate modified for testing)